### PR TITLE
Add Target to Extract Binary Artifacts from RPMs

### PR DESCRIPTION
**Pull request by adamperlin**
**What this PR does / why we need it**:
There is a desire from some Dalec users to extract the binary artifacts produced as part of Dalec's rpm builds as standalone files.  This PR adds a target which produces a single zip file containing the binary artifacts specified in a Dalec spec. Only the binaries are included with no dependencies whatsoever, so it is the user's responsibility to ensure that any standalone binaries extracted with this target are fully statically linked. 

**Which issue(s) this PR fixes** *(optional, using `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when the PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by adamperlin** [comment]
Haven't determined the source of the CI failures yet, it appears to be some kind of timeout

**Comment by cpuguy83** [comment]
Rebased to include #264

**Comment by cpuguy83** [comment]
Rebased this on top of #280 and tests are passing now.

**Comment by adamperlin** [comment]
> Rebased this on top of #280 and tests are passing now.

Thank you!!

File: frontend/azlinux/handle_bin.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+	return sb.String()
+}
+
+func zip(worker llb.State, zipName string, outputDir string, artifacts llb.State) llb.State {
+	outName := filepath.Join(outputDir, zipName+".zip")
+	return worker.Run(
+		shArgs("zip "+outName+" *"),
+		llb.Dir("/tmp/artifacts"),
+		llb.AddMount("/tmp/artifacts", artifacts)).AddMount(outputDir, llb.Scratch())
+}
```
Comment (DannyBrito) [inline frontend/azlinux/handle_bin.go:60]: Can't this `zip` function be deleted and use directly the `Zip` from `dalec` package in the `helpers.go` file? These seem to be the same, no?

Comment (adamperlin) [reply frontend/azlinux/handle_bin.go:60]: Ah yes thank you, I think I meant to delete this

File: frontend/azlinux/handle_bin.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+				llb.AddMount("/script", scriptState),
+				llb.AddMount("/package", rpmState),
+				llb.AddMount("/extracted", llb.Scratch()),
+			).AddMount("/out", llb.Scratch())
+
+			pg = dalec.ProgressGroup("Compressing artifacts: ")
+
+			zipWorker := w.Base(client, pg).Run(w.Install([]string{"zip"})).Root()
+
+			zipped := zip(zipWorker, "binaries", "/out", st)
```
Comment (adamperlin) [inline frontend/azlinux/handle_bin.go:99]: Would we prefer `.zip` or `.tar.gz`?

File: test/bin_extract_test.go [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+	"os"
+	"testing"
+
+	"github.com/Azure/dalec"
+	"github.com/Azure/dalec/frontend/pkg/bkfs"
+	gwclient "github.com/moby/buildkit/frontend/gateway/client"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestBinExtract_Mariner2(t *testing.T) {
```
Comment (cpuguy83) [inline test/bin_extract_test.go:19]: We should be able to move these tests directly into the distro test.

Comment (cpuguy83) [reply test/bin_extract_test.go:19]: Reason being this makes sure when we add new targets we don't forget to add these extra tests.

File: test/bin_extract_test.go [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+	if err != nil {
+		t.Fatalf("unable to read zip: %s", err)
+	}
+
+	zr, err := zip.NewReader(bytes.NewReader(read), info.Size())
+	if err != nil {
+		t.Fatal("unable to create zip reader %w", err)
+	}
+
+	for _, file := range zr.File {
```
Comment (cpuguy83) [inline test/bin_extract_test.go:52]: All the assertions in here should include the file name.

File: test/bin_extract_test.go [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+			Artifacts: dalec.Artifacts{
+				Binaries: map[string]dalec.ArtifactConfig{
+					"src/phony.sh": {},
+				},
+			},
+		}
+
+		testEnv.RunTest(ctx, t, func(ctx context.Context, client gwclient.Client) (*gwclient.Result, error) {
+			sr := newSolveRequest(withBuildTarget(buildTarget), withSpec(ctx, t, spec))
+			sr.Evaluate = true
```
Comment (cpuguy83) [inline test/bin_extract_test.go:100]: IIRC evaluate is being set to true by default now.

File: test/bin_extract_test.go [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+	if err != nil {
+		t.Fatalf("unable to stat zip: %s", err)
+	}
+
+	read, err := io.ReadAll(zf)
+	if err != nil {
+		t.Fatalf("unable to read zip: %s", err)
+	}
+
+	zr, err := zip.NewReader(bytes.NewReader(read), info.Size())
```
Comment (cpuguy83) [inline test/bin_extract_test.go:47]: Should be able to drop the `fs.File` into `zip.NewReader` directly.

File: test/bin_extract_test.go [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+type expectFile struct {
+	contents    string
+	permissions os.FileMode
+}
+
+func assertZipContentsMatch(t *testing.T, zf fs.File, want map[string]expectFile) {
+	info, err := zf.Stat()
+	if err != nil {
+		t.Fatalf("unable to stat zip: %s", err)
+	}
```
Comment (cpuguy83) [inline test/bin_extract_test.go:40]:
```suggestion
	assert.NilError(t, err)
```

File: test/bin_extract_test.go [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+
+	zr, err := zip.NewReader(bytes.NewReader(read), info.Size())
+	if err != nil {
+		t.Fatal("unable to create zip reader %w", err)
+	}
+
+	for _, file := range zr.File {
+		f, err := file.Open()
+		assert.Nil(t, err)
```
Comment (cpuguy83) [inline test/bin_extract_test.go:54]:
```suggestion
		assert.NilError(t, err)
```

File: test/bin_extract_test.go [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+		contents, err := io.ReadAll(f)
+
+		assert.Nil(t, err)
```
Comment (cpuguy83) [inline test/bin_extract_test.go:57]:
```suggestion
		assert.NilError(t, err)
```

File: test/bin_extract_test.go [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+
+	read, err := io.ReadAll(zf)
+	if err != nil {
+		t.Fatalf("unable to read zip: %s", err)
+	}
+
+	zr, err := zip.NewReader(bytes.NewReader(read), info.Size())
+	if err != nil {
+		t.Fatal("unable to create zip reader %w", err)
+	}
```
Comment (cpuguy83) [inline test/bin_extract_test.go:50]:
```suggestion
	assert.NilError(t, err)
```

File: frontend/azlinux/handle_bin.go (hunk 2 above)
Comment (cpuguy83) [reply frontend/azlinux/handle_bin.go:99]: I don't think we need to do anything with the binaries, we can just keep them in a raw llb.State.
Same for Windows.

**Review by cpuguy83 — CHANGES_REQUESTED** [review]

File: frontend/azlinux/handler.go [hunk 11]
```diff
@@ -33,6 +33,11 @@ func newHandler(w worker) gwclient.BuildFunc {
 	})
 	mux.Add("rpm/debug", handleDebug(w), nil)
 
+	mux.Add("rpm/bin", handleBin(w), &targets.Target{
```
[the remark this answers is not in the document]
Comment (adamperlin) [reply frontend/azlinux/handler.go:36]: Just to clarify are you suggesting renaming `windows/zip` to `windows/artifacts/binaries` then?

[the remark this answers is not in the document]
Comment (cpuguy83) [reply frontend/azlinux/handler.go:36]: I am suggesting to add a new target `windows/artifacts/binaries`

File: test/bin_extract_test.go (hunk 9 above)
Comment (adamperlin) [reply test/bin_extract_test.go:57]: I don't think that method exists with the testify library

Comment (cpuguy83) [reply test/bin_extract_test.go]: Ah that explains it, we have 2 assertion libraries in the repo.

Comment (adamperlin) [reply test/bin_extract_test.go]: Oh interesting. Perhaps we should pare it down to one?

File: frontend/azlinux/handle_bin.go [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+	sb.WriteString(`
+#!/bin/bash
+set -e
+declare -a RPMS=()
+export RPM_BINDIR=$(rpm --eval '%{_bindir}')
+`)
+
+	for _, rpm := range rpms {
+		sb.WriteString(fmt.Sprintf("RPMS+=(%q)\n", rpm))
+	}
```
Comment (pmengelbert) [inline frontend/azlinux/handle_bin.go:29]: Bash has built-in ways of building up arrays from text. My personal choice is to use [readarray](https://helpmanual.io/builtin/readarray/) (an alias of `mapfile`). You have to be sure you're using bash and not some other shell, and you have to be sure that your bash version is >= 4 (almost all distros these days).

You could get the data into the container by
```go
rpmsData := []byte(strings.Join(rpms, "\x00")) // null separate the rpms
```

Then, you can take those bytes and use `llb.MkFile` to mount it into the container, for example at `/tmp/rpm_list`. In bash, you would do:

```bash
readarray -d '' RPMS < /tmp/rpm_list # reads the file and puts the results into the array variable "$RPMS"
```

This note is for educational purposes, what you have is fine.

File: frontend/azlinux/handle_bin.go [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+)
+
+func binCopyScript(rpms []string, binaries map[string]dalec.ArtifactConfig) string {
+	var sb strings.Builder
+
+	sb.WriteString(`
+#!/bin/bash
+set -e
+declare -a RPMS=()
+export RPM_BINDIR=$(rpm --eval '%{_bindir}')
```
[the remark this answers is not in the document]
Comment (adamperlin) [reply frontend/azlinux/handle_bin.go:24]: So just to confirm, are we ok with having this just extract from `%{_bindir}` (and any subpaths) or do we want to look in any additional locations?

File: frontend/azlinux/handle_bin.go (hunk 12 above)
Comment (adamperlin) [reply frontend/azlinux/handle_bin.go:29]: Thanks for this, I am not super familiar with bash array building capabilities, so I appreciate it!

File: frontend/azlinux/handle_bin.go [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+
+	for _, rpm := range rpms {
+		sb.WriteString(fmt.Sprintf("RPMS+=(%q)\n", rpm))
+	}
+
+	sb.WriteString("for rpm in $RPMS; do\n")
+	binaryPathList := make([]string, 0, len(binaries))
+	for path, bin := range binaries {
+		srcPath := bin.InstallPath(path)
+		binaryPathList = append(binaryPathList, filepath.Join(".$RPM_BINDIR", srcPath))
```
[the remark this answers is not in the document]
Comment (adamperlin) [reply frontend/azlinux/handle_bin.go]: Changed this. I discovered that `filepath.Join` calls  `filepath.Clean` which will strip off a leading `./`, so the `./` must be appended after the call

File: frontend/azlinux/handle_bin.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (pmengelbert) [reply frontend/azlinux/handle_bin.go]: > So just to confirm, are we ok with having this just extract from `%{_bindir}`

Yes, unless I've misunderstood.

**Review by pmengelbert — APPROVED** [review]
LGTM, great work on this

File: frontend/azlinux/handle_bin.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (cpuguy83) [reply frontend/azlinux/handle_bin.go]: Not ignoring this just need some time to think about the edge cases.

**Comment by adamperlin** [comment]
Thanks for the review and sign off @pmengelbert! After a bit of offline discussion today, I think we've decided to put this PR on hold for now as it may not actually be needed to serve the use case

**Comment by cpuguy83** [comment]
We discussed this and are thinking this should be handled outside of dalec.
It is difficult to solve all use cases with this and do not want to encourage people to extract the binaries (which are typically going to be dynamically linked anyway).

**Comment by adamperlin** [comment]
Closing this due to above discussion.

**Pull request closed without merging**